helix-event/src/debounce.rs

```diff
@@ -0,0 +1,62 @@
+//! Utilities for declaring an async (usually debounced) hook
+
+use std::time::Duration;
+
+use futures_executor::block_on;
+use tokio::sync::mpsc::{self, error::TrySendError, Sender};
+use tokio::time::Instant;
+
+/// Async hooks are the core of the event system, the m
+pub trait AsyncHook: Sync + Send + 'static + Sized {
+    type Event: Sync + Send + 'static;
+    /// Called immidietly whenever an event is received, this function can
+    /// consume the event immidietly or debounce it. In case of debouncing
+    /// it can either define a new debounce timeout or continue the current
+    fn handle_event(&mut self, event: Self::Event, timeout: Option<Instant>) -> Option<Instant>;
+
+    /// Called whenever the debounce timeline is searched
+    fn finish_debounce(&mut self);
+
+    fn spawn(self) -> mpsc::Sender<Self::Event> {
```

Comment:
hmm I don't really see as much point in this case, the cancelation case is a type alias because at the end of the day, the user doesn't really need to know its a channel internally but in this case you want to know its a channel.

I man not entirely opposed to it. `type AsyncHandle<Handler: AsyncHook> = mpsc::Sender<Handler::Event>` could be interesting. But on the other hand I think `Sender<CompletionEvent>` is just as clearer than `AsyncHandle<CompletinoHandler>`. Like it look/sounds cool but it doesn't really seem to improve code clarity.

Gonna need to sleep on this one :D

Reply:
Yeah, just a thought I had, I don't feel strongly about it at all.

Reply:
I think I would prefer the current signature here. I think `Sender<CompletionEvent>` is quite clear